math/Line2.cs

```diff
@@ -1,51 +1,50 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
 
 namespace g3
 {
-    public struct Line2d
+    public readonly struct Line2d
     {
-        public Vector2d Origin;
-        public Vector2d Direction;
+        public readonly Vector2d Origin;
+        public readonly Vector2d Direction;
 
-        public Line2d(Vector2d origin, Vector2d direction) {
-            this.Origin = origin;
-            this.Direction = direction;
+        public Line2d(Vector2d origin, Vector2d direction)
+        {
+            Origin = origin;
+            Direction = direction;
         }
 
-        public Line2d(ref Vector2d origin, ref Vector2d direction) {
-            this.Origin = origin;
-            this.Direction = direction;
+        public Line2d(ref Vector2d origin, ref Vector2d direction)
+        {
+            Origin = origin;
+            Direction = direction;
         }
 
-        public static Line2d FromPoints(Vector2d p0, Vector2d p1) {
-            return new Line2d(p0, (p1 - p0).Normalized);
-        }
-        public static Line2d FromPoints(ref Vector2d p0, ref Vector2d p1) {
-            return new Line2d(p0, (p1 - p0).Normalized);
-        }
+        public static Line2d FromPoints(Vector2d p0, Vector2d p1) =>
+            new Line2d(p0, (p1 - p0).Normalized);
+
+        public static Line2d FromPoints(ref Vector2d p0, ref Vector2d p1) =>
+            new Line2d(p0, (p1 - p0).Normalized);
 
         // parameter is distance along Line
-        public Vector2d PointAt(double d) {
-            return Origin + d * Direction;
-        }
+        public Vector2d PointAt(double d) =>
+            Origin + d * Direction;
+
+        public double Project(in Vector2d p) =>
+            (p - Origin).Dot(Direction);
 
-        public double Project(Vector2d p)
+        public Vector2d ClosestPoint(in Vector2d p)
         {
-            return (p - Origin).Dot(Direction);
+            double t = (p - Origin).Dot(Direction);
+            return Origin + t * Direction;
         }
 
-        public double DistanceSquared(Vector2d p)
+        public double DistanceSquared(in Vector2d p)
         {
             double t = (p - Origin).Dot(Direction);
             Vector2d proj = Origin + t * Direction;
```

Comment:
these two lines are equivalent to ClosestPoint method

Reply:
They are but g3 authors (and we follow their example) rarely re-use method calls as these primitive types can be used in performance critical context.

Reply:
alright, if this is crucial for performance, I am fine with that